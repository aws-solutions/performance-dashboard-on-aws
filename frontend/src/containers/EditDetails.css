.edit-details-form {
  max-width: 100% !important;
}
  
.edit-details-form-group-topicarea-id {
  width: 75%;
}
  
.edit-details-form-group-dashboard-name {
  width: 50%;
}