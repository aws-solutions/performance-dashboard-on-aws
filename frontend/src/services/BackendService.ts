/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import API from "@aws-amplify/api";
import Auth from "@aws-amplify/auth";
import { OutgoingHttpHeaders } from "http2";
import {
    Dashboard,
    DashboardVersion,
    Dataset,
    PublicDashboard,
    PublicHomepage,
    Widget,
    User,
    DatasetSchema,
    DashboardAuditLog,
} from "../models";

const apiName = "BackendApi";
const publicPath = "public";

async function authHeaders() {
    const token = await getAuthToken();
    return {
        Authorization: "Bearer ".concat(token),
    };
}

async function fetchCsrfToken(): Promise<string> {
    let headers = {};
    if (window.EnvironmentConfig?.authenticationRequired) {
        headers = await authHeaders();
    }
    const data = await API.get(apiName, `${publicPath}/csrf-token`, {
        headers,
        withCredentials: true,
    });
    return data.token;
}

async function addCsrfToken(headers: OutgoingHttpHeaders) {
    const token = await fetchCsrfToken();
    headers["x-csrf-token"] = token;
    return headers;
}

async function getAuthToken() {
    const session = await Auth.currentSession();
    const idToken = session.getIdToken();
    return idToken.getJwtToken();
}

async function fetchDashboards(): Promise<Array<Dashboard>> {
    const headers = await authHeaders();
    return API.get(apiName, "dashboard", { headers });
}

async function fetchDashboardVersions(parentDashboardId: string): Promise<Array<DashboardVersion>> {
    const headers = await authHeaders();
    return API.get(apiName, `dashboard/${parentDashboardId}/versions`, {
        headers,
    });
}

async function fetchDashboardHistory(parentDashboardId: string): Promise<Array<DashboardAuditLog>> {
    const headers = await authHeaders();
    return API.get(apiName, `dashboard/${parentDashboardId}/auditlogs`, {
        headers,
    });
}

async function fetchDashboardById(dashboardId: string): Promise<Dashboard> {
    const headers = await authHeaders();
    return API.get(apiName, `dashboard/${dashboardId}`, { headers });
}

async function fetchPublicDashboardByURL(friendlyURL: string): Promise<Dashboard> {
    let headers = {};
    if (window.EnvironmentConfig?.authenticationRequired) {
        headers = await authHeaders();
    }
    return API.get(apiName, `${publicPath}/dashboard/friendly-url/${friendlyURL}`, { headers });
}

async function fetchPublicHomepageWithQuery(query: string): Promise<PublicHomepage> {
    let headers = {};
    if (window.EnvironmentConfig?.authenticationRequired) {
        headers = await authHeaders();
    }
    if (query === undefined || query === "") {
        return API.get(apiName, `${publicPath}/homepage`, { headers });
    } else {
        return API.get(apiName, `${publicPath}/search?q=${query}`, {
            headers,
        });
    }
}

async function fetchTopicAreas() {
    const headers = await authHeaders();
    return API.get(apiName, "topicarea", { headers });
}

async function fetchTopicAreaById(topicAreaId: string) {
    const headers = await authHeaders();
    return API.get(apiName, `topicarea/${topicAreaId}`, { headers });
}

async function fetchWidgetById(dashboardId: string, widgetId: string): Promise<Widget> {
    const headers = await authHeaders();
    return API.get(apiName, `dashboard/${dashboardId}/widget/${widgetId}`, {
        headers,
    });
}

async function fetchWidgets(dashboardId: string) {
    const headers = await authHeaders();
    return API.get(apiName, `dashboard/${dashboardId}/widgets`, {
        headers,
    });
}

async function createDashboard(name: string, topicAreaId: string, description: string) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.post(apiName, "dashboard", {
        headers,
        withCredentials: true,
        body: {
            name,
            topicAreaId,
            description,
        },
    });
}

async function editDashboard(
    dashboardId: string,
    name: string,
    topicAreaId: string,
    displayTableOfContents: boolean,
    description: string,
    updatedAt: Date,
    tableOfContents?: any,
) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, `dashboard/${dashboardId}`, {
        headers,
        withCredentials: true,
        body: {
            name,
            topicAreaId,
            displayTableOfContents,
            tableOfContents,
            description,
            updatedAt,
        },
    });
}

async function publishDashboard(
    dashboardId: string,
    updatedAt: Date,
    releaseNotes: string,
    friendlyURL?: string,
) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, `dashboard/${dashboardId}/publish`, {
        headers,
        withCredentials: true,
        body: {
            updatedAt,
            releaseNotes,
            friendlyURL,
        },
    });
}

async function deleteDashboards(dashboards: Array<string>) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.del(apiName, `dashboard?ids=${dashboards.join(",")}`, {
        headers,
        withCredentials: true,
    });
}

async function createTopicArea(name: string) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.post(apiName, "topicarea", {
        headers,
        withCredentials: true,
        body: {
            name,
        },
    });
}

async function renameTopicArea(topicAreaId: string, name: string) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, `topicarea/${topicAreaId}`, {
        headers,
        withCredentials: true,
        body: {
            name,
        },
    });
}

async function deleteTopicArea(topicareaId: string) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.del(apiName, `topicarea/${topicareaId}`, {
        headers,
        withCredentials: true,
    });
}

async function archive(dashboardId: string, lastUpdatedAt: Date): Promise<Dashboard> {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, `dashboard/${dashboardId}/archive`, {
        headers,
        withCredentials: true,
        body: {
            updatedAt: lastUpdatedAt,
        },
    });
}

async function createWidget(
    dashboardId: string,
    name: string,
    widgetType: string,
    showTitle: boolean,
    content: object,
) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.post(apiName, `dashboard/${dashboardId}/widget`, {
        headers,
        withCredentials: true,
        body: {
            name,
            widgetType,
            showTitle,
            content,
        },
    });
}

async function editWidget(
    dashboardId: string,
    widgetId: string,
    name: string,
    showTitle: boolean,
    content: object,
    updatedAt: Date,
) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, `dashboard/${dashboardId}/widget/${widgetId}`, {
        headers,
        withCredentials: true,
        body: {
            name,
            showTitle,
            content,
            updatedAt,
        },
    });
}

async function duplicateWidget(
    dashboardId: string,
    widgetId: string,
    updatedAt: Date,
    copyLabel: string,
) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.post(apiName, `dashboard/${dashboardId}/widget/${widgetId}`, {
        headers,
        withCredentials: true,
        body: {
            updatedAt,
            copyLabel,
        },
    });
}

async function deleteWidget(dashboardId: string, widgetId: string) {
    const headers = await authHeaders();
    const url = `dashboard/${dashboardId}/widget/${widgetId}`;
    await addCsrfToken(headers);
    return API.del(apiName, url, {
        headers,
        withCredentials: true,
    });
}

async function setWidgetOrder(dashboardId: string, widgets: Array<Widget>): Promise<Dataset> {
    const headers = await authHeaders();
    const payload = widgets.map((widget) => ({
        id: widget.id,
        updatedAt: widget.updatedAt,
        order: widget.order,
        section: widget.section,
        content: widget.content,
    }));
    await addCsrfToken(headers);
    return API.put(apiName, `dashboard/${dashboardId}/widgetorder`, {
        headers,
        withCredentials: true,
        body: {
            widgets: payload,
        },
    });
}

async function fetchDatasets(): Promise<Array<Dataset>> {
    const headers = await authHeaders();
    return API.get(apiName, "dataset", { headers });
}

async function createDataset(
    fileName: string,
    s3Keys: { raw: string; json: string },
    schema = DatasetSchema.None,
): Promise<Dataset> {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.post(apiName, "dataset", {
        headers,
        withCredentials: true,
        body: {
            fileName,
            s3Key: {
                raw: s3Keys.raw,
                json: s3Keys.json,
            },
            schema,
        },
    });
}

async function fetchHomepage() {
    const headers = await authHeaders();
    return API.get(apiName, "settings/homepage", { headers });
}

async function fetchPublicHomepage() {
    let headers = {};
    if (window.EnvironmentConfig?.authenticationRequired) {
        headers = await authHeaders();
    }
    return API.get(apiName, `${publicPath}/homepage`, { headers });
}

async function editHomepage(title: string, description: string, updatedAt: Date) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, "settings/homepage", {
        headers,
        withCredentials: true,
        body: {
            title,
            description,
            updatedAt,
        },
    });
}

async function fetchSettings() {
    const headers = await authHeaders();
    return API.get(apiName, "settings", { headers });
}

async function fetchPublicSettings() {
    let headers = {};
    if (window.EnvironmentConfig?.authenticationRequired) {
        headers = await authHeaders();
    }
    return API.get(apiName, `${publicPath}/settings`, { headers });
}

async function editSettings(publishingGuidance: string, updatedAt: Date) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, "settings", {
        headers,
        withCredentials: true,
        body: {
            publishingGuidance,
            updatedAt,
        },
    });
}

async function updateSetting(settingKey: string, settingValue: any, updatedAt: Date) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, "settings", {
        headers,
        withCredentials: true,
        body: {
            [settingKey]: settingValue,
            updatedAt,
        },
    });
}

async function fetchPublicDashboard(dashboardId: string): Promise<PublicDashboard> {
    let headers = {};
    if (window.EnvironmentConfig?.authenticationRequired) {
        headers = await authHeaders();
    }
    return API.get(apiName, `${publicPath}/dashboard/${dashboardId}`, {
        headers,
    });
}

async function createDraft(dashboardId: string): Promise<Dashboard> {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.post(apiName, `dashboard/${dashboardId}`, {
        headers,
        withCredentials: true,
    });
}

async function publishPending(
    dashboardId: string,
    lastUpdatedAt: Date,
    releaseNotes?: string,
): Promise<Dashboard> {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, `dashboard/${dashboardId}/publishpending`, {
        headers,
        withCredentials: true,
        body: {
            updatedAt: lastUpdatedAt,
            releaseNotes,
        },
    });
}

async function moveToDraft(dashboardId: string, lastUpdatedAt: Date): Promise<Dashboard> {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, `dashboard/${dashboardId}/draft`, {
        headers,
        withCredentials: true,
        body: {
            updatedAt: lastUpdatedAt,
        },
    });
}

async function fetchUsers(): Promise<User[]> {
    const headers = await authHeaders();
    return API.get(apiName, "user", { headers });
}

async function addUsers(role: string, emails: Array<string>) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.post(apiName, "user", {
        headers,
        withCredentials: true,
        body: {
            role,
            emails: emails.join(","),
        },
    });
}

async function removeUsers(usernames: Array<string>) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.del(apiName, "user", {
        headers,
        withCredentials: true,
        body: {
            usernames,
        },
    });
}

async function resendInvite(emails: Array<string>) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.post(apiName, "user/invite", {
        headers,
        withCredentials: true,
        body: {
            emails: emails.join(","),
        },
    });
}

async function changeRole(role: string, usernames: Array<string>) {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.put(apiName, "user/role", {
        headers,
        withCredentials: true,
        body: {
            role,
            usernames,
        },
    });
}

async function copyDashboard(dashboardId: string): Promise<Dashboard> {
    const headers = await authHeaders();
    await addCsrfToken(headers);
    return API.post(apiName, `dashboard/${dashboardId}/copy`, { headers, withCredentials: true });
}

const BackendService = {
    fetchDashboards,
    fetchDashboardById,
    fetchPublicHomepageWithQuery,
    fetchTopicAreas,
    fetchTopicAreaById,
    fetchWidgetById,
    fetchWidgets,
    editDashboard,
    publishDashboard,
    createDashboard,
    deleteDashboards,
    createTopicArea,
    renameTopicArea,
    deleteTopicArea,
    createWidget,
    editWidget,
    duplicateWidget,
    deleteWidget,
    setWidgetOrder,
    createDataset,
    fetchDatasets,
    getAuthToken,
    fetchHomepage,
    fetchPublicHomepage,
    editHomepage,
    fetchPublicDashboard,
    fetchPublicDashboardByURL,
    fetchSettings,
    fetchPublicSettings,
    editSettings,
    updateSetting,
    publishPending,
    archive,
    createDraft,
    moveToDraft,
    fetchDashboardVersions,
    fetchDashboardHistory,
    fetchUsers,
    addUsers,
    removeUsers,
    resendInvite,
    changeRole,
    copyDashboard,
    fetchCsrfToken,
};

export default BackendService;
