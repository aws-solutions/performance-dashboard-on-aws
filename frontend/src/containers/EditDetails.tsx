import React, { useEffect, useState } from "react";
import { useHistory, useParams } from "react-router-dom";
import { useForm } from "react-hook-form";
import { useTopicAreas, useDashboard, useSettings } from "../hooks";
import BackendService from "../services/BackendService";
import Markdown from "../components/Markdown";
import TextField from "../components/TextField";
import Dropdown from "../components/Dropdown";
import Button from "../components/Button";
import Breadcrumbs from "../components/Breadcrumbs";
import Spinner from "../components/Spinner";
import DashboardHeader from "../components/DashboardHeader";
import Link from "../components/Link";

interface FormValues {
  name: string;
  topicAreaId: string;
  description: string;
}

interface PathParams {
  dashboardId: string;
}

function EditDetails() {
  const history = useHistory();
  const { settings } = useSettings();
  const { topicareas } = useTopicAreas();
  const { dashboardId } = useParams<PathParams>();
  const { dashboard, loading } = useDashboard(dashboardId);
  const { register, errors, handleSubmit, getValues } = useForm<FormValues>();
  const [name, setName] = useState("");
  const [topicAreaName, setTopicAreaName] = useState("");
  const [description, setDescription] = useState("");

  useEffect(() => {
    if (dashboard) {
      setName(dashboard.name);
      setTopicAreaName(dashboard.topicAreaName);
      setDescription(dashboard.description || "");
    }
  }, [dashboard]);

  const onSubmit = async (values: FormValues) => {
    await BackendService.editDashboard(
      dashboardId,
      values.name,
      values.topicAreaId,
      values.description || "",
      dashboard ? dashboard.updatedAt : new Date()
    );

    history.push(`/admin/dashboard/edit/${dashboardId}`, {
      alert: {
        type: "success",
        message: `"${values.name}" details successfully edited`,
      },
      id: "top-alert",
    });
  };

  const onFormChange = () => {
    const { name, topicAreaId, description } = getValues();
    setName(name);
    setTopicAreaName(topicareas.find((t) => t.id === topicAreaId)?.name || "");
    setDescription(description);
  };

  const onCancel = () => {
    history.push(`/admin/dashboard/edit/${dashboardId}`);
  };

  const crumbs = [
    {
      label: "Dashboards",
      url: "/admin/dashboards",
    },
    {
      label: dashboard?.name,
      url: `/admin/dashboard/edit/${dashboard?.id}`,
    },
  ];

  if (!loading) {
    crumbs.push({
      label: "Edit details",
      url: "",
    });
  }

  return (
    <>
      <Breadcrumbs crumbs={crumbs} />
      <h1>Edit Details</h1>

      {loading || !dashboard || !topicareas || topicareas.length === 0 ? (
        <Spinner className="text-center margin-top-9" label="Loading" />
      ) : (
        <div className="grid-row width-desktop">
          <div className="grid-col-6">
            <div className="grid-row">
              <div className="grid-col-12">
                <form
                  onSubmit={handleSubmit(onSubmit)}
                  className="edit-details-form usa-form usa-form--large"
                  onChange={onFormChange}
                  data-testid="EditDetailsForm"
                >
                  <TextField
                    id="name"
                    name="name"
                    label="Dashboard Name"
                    error={errors.name && "Please specify a name"}
                    defaultValue={dashboard?.name}
                    register={register}
                    required
                  />

                  <Dropdown
                    id="topicAreaId"
                    name="topicAreaId"
                    label={settings.topicAreaLabels.singular}
                    hint={`Select an existing ${settings.topicAreaLabels.singular.toLowerCase()}`}
                    defaultValue={dashboard?.topicAreaId}
                    register={register}
                    options={topicareas.map((topicarea) => ({
                      value: topicarea.id,
                      label: topicarea.name,
                    }))}
                  />

                  <TextField
                    id="description"
                    name="description"
                    label="Description - optional"
                    defaultValue={dashboard?.description}
                    hint={
                      <>
                        Give your dashboard a description to explain it in more
                        depth. This text area supports limited markdown.{" "}
                        <Link target="_blank" to={"/admin/markdown"} external>
                          View Markdown Syntax
                        </Link>
                      </>
                    }
                    register={register}
                    multiline
                    rows={10}
                  />

                  <br />
                  <Button type="submit" disabled={loading}>
                    Save
                  </Button>
                  <Button
                    variant="unstyled"
                    type="button"
                    className="margin-left-1 text-base-dark hover:text-base-darker active:text-base-darkest"
                    onClick={onCancel}
                  >
                    Cancel
                  </Button>
                </form>
              </div>
            </div>
          </div>
          <div className="grid-col-6">
            <div className="margin-left-3 margin-top-2">
              <DashboardHeader
                name={name}
                topicAreaName={topicAreaName}
                description={description}
              />
              <hr
                style={{
                  border: "none",
                  height: "1px",
                  backgroundColor: "#dfe1e2",
                  margin: "2rem 0",
                }}
              />
            </div>
          </div>
        </div>
      )}
    </>
  );
}

export default EditDetails;
