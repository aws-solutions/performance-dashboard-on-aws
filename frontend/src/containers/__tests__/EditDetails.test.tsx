/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render, fireEvent, act, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createMemoryHistory } from "history";
import { Route, Router } from "react-router-dom";
import BackendService from "../../services/BackendService";
import EditDetails from "../EditDetails";

jest.mock("../../hooks");
jest.mock("../../services/BackendService");

const history = createMemoryHistory();
history.push("/admin/dashboard/edit/123/header");
jest.spyOn(history, "push");

beforeEach(async () => {
    await act(async () => {
        render(
            <Router history={history}>
                <Route path="/admin/dashboard/edit/:dashboardId/header">
                    <EditDetails />
                </Route>
            </Router>,
        );
    });
});

test("submits form with the entered values", async () => {
    userEvent.selectOptions(screen.getByLabelText("Ministry"), "123456789");

    await act(async () => {
        fireEvent.submit(screen.getByTestId("EditDetailsForm"));
    });

    expect(BackendService.editDashboard).toBeCalledWith(
        "123",
        "My AWS Dashboard",
        "123456789",
        false,
        "Some description",
        "",
        {},
    );
});

test("invokes cancel function when use clicks cancel", async () => {
    await act(async () => {
        fireEvent.click(screen.getByRole("button", { name: "Cancel" }));
    });
    expect(history.push).toHaveBeenCalledWith("/admin/dashboard/edit/123");
});

test("renders a preview of dashboard name and description", async () => {
    fireEvent.input(screen.getByLabelText("Dashboard Name"), {
        target: {
            value: "Foo Bar",
        },
    });

    fireEvent.input(screen.getByLabelText("Description (optional)"), {
        target: {
            value: "FizzBuzz",
        },
    });

    const description = screen.getByText("FizzBuzz");
    const name = screen.getByRole("heading", { name: "Foo Bar" });

    await waitFor(() => expect(name).toBeInTheDocument());
    await waitFor(() => expect(description).toBeInTheDocument());
});
