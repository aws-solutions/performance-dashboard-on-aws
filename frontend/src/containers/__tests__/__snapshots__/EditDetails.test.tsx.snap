// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit details should match snapshot 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    class="usa-breadcrumb padding-top-0"
  >
    <ol
      class="usa-breadcrumb__list"
    >
      <li
        class="usa-breadcrumb__list-item"
      >
        <a
          aria-label=""
          class="link"
          href="/admin/dashboards"
          target=""
        >
          <span>
            <span>
              Dashboards
            </span>
          </span>
        </a>
      </li>
      <li
        class="usa-breadcrumb__list-item"
      >
        <a
          aria-label=""
          class="link"
          href="/admin/dashboard/edit/123"
          target=""
        >
          <span>
            <span>
              My AWS Dashboard
            </span>
          </span>
        </a>
      </li>
      <li
        aria-current="page"
        class="usa-breadcrumb__list-item usa-current cursor-default"
      >
        <span>
          Edit details
        </span>
      </li>
    </ol>
  </nav>
  <h1>
    Edit Details
  </h1>
  <div
    class="grid-row width-desktop"
  >
    <div
      class="grid-col-6"
    >
      <div
        class="grid-row"
      >
        <div
          class="grid-col-12"
        >
          <form
            class="edit-details-form usa-form usa-form--large"
            data-testid="EditDetailsForm"
          >
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label text-bold"
                for="name"
              >
                Dashboard Name
              </label>
              <div
                class="usa-hint"
              />
              <input
                class="usa-input"
                id="name"
                name="name"
                type="text"
                value="My AWS Dashboard"
              />
            </div>
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label text-bold"
                for="topicAreaId"
              >
                Ministry
              </label>
              <div
                class="usa-hint"
                id="event-date-start-hint"
              >
                Select an existing ministry
              </div>
              <select
                class="usa-select"
                id="topicAreaId"
                name="topicAreaId"
              >
                <option
                  value="123456789"
                >
                  Topic Area Bananas
                </option>
                <option
                  value="987654321"
                >
                  Topic Area Grapes
                </option>
              </select>
            </div>
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label text-bold"
                for="description"
              >
                Description - optional
              </label>
              <div
                class="usa-hint"
              >
                Give your dashboard a description to explain it in more depth. This text area supports limited markdown.
                 
                <a
                  aria-label=""
                  class="link"
                  href="/admin/markdown"
                  target="_blank"
                >
                  <span>
                    View Markdown Syntax
                  </span>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-external-link-alt fa-w-16 fa-xs margin-left-05"
                    data-icon="external-link-alt"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                      fill="currentColor"
                    />
                  </svg>
                </a>
              </div>
              <textarea
                class="usa-textarea"
                id="description"
                name="description"
                rows="10"
                style="height: auto;"
              >
                Some description
              </textarea>
            </div>
            <br />
            <button
              class="usa-button usa-button--base"
              type="submit"
            >
              Save
            </button>
            <button
              class="usa-button usa-button--unstyled margin-left-1 text-base-dark hover:text-base-darker active:text-base-darkest"
              type="button"
            >
              Cancel
            </button>
          </form>
        </div>
      </div>
    </div>
    <div
      class="grid-col-6"
    >
      <div
        class="margin-left-3 margin-top-2"
      >
        <div
          class="margin-top-2"
        >
          <h1
            class="display-inline-block margin-bottom-1 font-sans-2xl"
          >
            My AWS Dashboard
          </h1>
          <div
            class="text-base text-italic margin-bottom-2"
          >
            Bananas
          </div>
        </div>
        <div
          class="margin-y-2"
        >
          <div
            class="Markdown font-sans-lg usa-prose margin-top-0"
          >
            <p>
              Some description
            </p>
          </div>
        </div>
        <hr
          style="height: 1px; background-color: rgb(223, 225, 226); margin: 2rem 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;
